src/core/components/pin.js

```diff
@@ -0,0 +1,589 @@
+'use strict'
+
+const dagPB = require('ipld-dag-pb')
+const DAGNode = dagPB.DAGNode
+const DAGLink = dagPB.DAGLink
+const CID = require('cids')
+const pinSet = require('./pin-set')
+const promisify = require('promisify-es6')
+const multihashes = require('multihashes')
+const Key = require('interface-datastore').Key
+const _ = require('lodash')
+
+const keyString = multihashes.toB58String
+function KeySet (keys) {
+  // Buffers with identical data are still different objects, so
+  // they need to be cast to strings to prevent duplicates in Sets
+  this.keys = {}
+  this.add = (key) => {
+    this.keys[keyString(key)] = key
+  }
+  this.delete = (key) => {
+    delete this.keys[keyString(key)]
+  }
+  this.clear = () => {
+    this.keys = {}
+  }
+  this.has = (key) => {
+    return (keyString(key) in this.keys)
+  }
+  this.toArray = () => {
+    return Object.keys(this.keys).map((hash) => {
+      return this.keys[hash]
+    })
+  }
+  this.toStringArray = () => {
+    return Object.keys(this.keys)
+  }
+  keys = keys || []
+  keys.forEach(this.add)
+}
+
+module.exports = function pin (self) {
+  let directPins = new KeySet()
+  let recursivePins = new KeySet()
+  let internalPins = new KeySet()
+
+  const pinDataStoreKey = new Key('/local/pins')
+
+  const repo = self._repo
+  const dag = self.dag
+
+  function normalizeHashes(hashes, callback) {
+    // try to accept a variety of hash options including
+    // multihash Buffers, base58 strings, and ipfs path
+    // strings, either individually or as an array
+    try {
+      if (Buffer.isBuffer(hashes) || typeof hashes.forEach !== 'function') {
+        hashes = [hashes]
+      }
+      const normalized = {
+        hashes: [],
+        update: (multihash) => {
+          multihashes.validate(multihash)
+          normalized.hashes.push(multihash)
+          if (hashes.length === normalized.hashes.length) {
+            return callback(null, normalized.hashes)
+          }
+        }
+      }
+      hashes.forEach((hash) => {
+        if (typeof hash === 'string') {
+          // example: '/ipfs/QmRootHash/links/by/name'
+          const matched = hash.match(/^(?:\/ipfs\/)?([^\/]+(?:\/[^\/]+)*)\/?$/)
+          if (!matched) {
+            return callback(new Error('invalid ipfs ref path'))
```

Comment:
this is calling a callback for each hash that is incorrect. Consider using async.each to test all the hashes and then return in case of error: https://caolan.github.io/async/docs.html#each